Move the implementation of Monitor to managed in CoreCLR

jkoritzinsky opened the pull request:
This allows us to share more code with NativeAOT, reduce a decent amount of complexity in the runtime, and fixes a blocking issue for https://github.com/dotnet/runtime/pull/117788

jkoritzinsky committed:
WIP: Move Monitor impl from unmanaged code to managed in CoreCLR

jkoritzinsky committed:
Clean up the build

jkoritzinsky committed:
Unify Condition implementation

jkoritzinsky committed:
Fix Mono and CoreCLR builds

jkoritzinsky committed:
Assert index for sync block

jkoritzinsky committed:
More ifdefs for mono

jkoritzinsky committed:
Add back ExitIfLockTaken for JIT helper

jkoritzinsky committed:
Fix GC mode

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Threading/SyncTable.CoreCLR.cs:
```diff
@@ -0,0 +1,34 @@
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "SyncTable_AssignEntry")]
+        private static partial int AssignEntryInternal(ObjectHandleOnStack obj);
```
This will likely need to have fast path FCall / slow path QCall split to avoid perf regressions.

jkoritzinsky committed:
Fix musl build and make gc modes explicit

jkoritzinsky committed:
Split GetLockObject's implementation into a fast FCall and a slow QCall

jkoritzinsky committed:
Fix some asserts

jkoritzinsky committed:
Lock uses 0-based recursion count. Also, only read the thin-lock member once to avoid racing.

Furthermore, handle an uninitialized lock ID in ObjectHeader.CoreCLR.cs. There's a possibility that the first time the managed Lock type's thread id tracking is used is on releasing a lock that was upgraded to a fat lock.

jkoritzinsky committed:
Fix musl build

jkoritzinsky committed:
FIx one more musl problem

jkoritzinsky committed:
Refactor out the object header spinlocking so we can use it on the thinlock for thin->fat lock upgrading

jkoritzinsky committed:
Only make the spinlock GC_NOTRIGGER for the object header and only go to COOP when creating the lock.

jkotas commented on src/coreclr/vm/comsynchronizable.cpp:
```diff
@@ -899,3 +892,49 @@ extern "C" void QCALLTYPE ThreadNative_ResetAbort()
+
+    _ASSERTE(0 <= idx && idx <= SyncBlockCache::GetSyncBlockCache()->GetTableEntryCount());
+
+    OBJECTHANDLE lock = SyncTableEntry::GetSyncTableEntry()[idx].m_SyncBlock->GetLock();
```
It is not safe to access `g_pSyncTable`  in cooperative mode. `g_pSyncTable` can grow and reclamation of the previous block is synchronized using GC suspension.

If you access `g_pSyncTable` in pre-emptive mode, you are at risk of accessing memory that was freed.

jkoritzinsky replied:
I'll move the sync block retrieval from the table back into coop mode.

jkoritzinsky committed:
Don't switch to the spinlock mechanism. Didn't solve the problem I thought it would and caused more problems.

jkoritzinsky committed:
Handle the case where an object is given a syncblock index and not had the syncblock created yet.

jkoritzinsky committed:
Don't allow a syncblock to be collected if it has thinlock data.

jkoritzinsky committed:
Reset thin-lock data on GC, give back an empty enumerable for GetBlockingObjects.

jkoritzinsky committed:
Fix arg passing on win-x86

jkoritzinsky committed:
Don't let our syncblock get cleaned up out from under us after SyncTable_AssignEntry

jkoritzinsky committed:
Send wait events from the managed side for monitor waits to avoid having to plumb through monitor waits back into CoreCLR

jkoritzinsky committed:
Respect the syncblock spin lock used in thin header->syncblock upgrade in our fast paths.

jkoritzinsky committed:
Update comment

jkoritzinsky committed:
Flip condition check

jkoritzinsky committed:
Update expected parameter name

jkoritzinsky committed:
Track the OS thread ID so diagnostics can pick it up

jkotas commented:
This needs a lot of perf testing of various scenarios. Have you done any perf testing on this?

jkoritzinsky commented:
I'm working on coming up with good benchmarks locally now. I'll try to build a good suite of ones and come back with results.

jkoritzinsky committed:
Don't track the OS thread ID, just use the managed thread ID

jkoritzinsky committed:
Force no-inlining for cases with P/Invokes (they're all a rare path)

jkoritzinsky committed:
Fix debug assert in dbi

jkoritzinsky commented:
I've done some perf testing and made some progress on optimization. Currently seeing the following performance for some simple local benchmarks. Definitely have a ways to go before this is ready to be merged.

```

BenchmarkDotNet v0.15.2, Windows 11 (10.0.26100.4946/24H2/2024Update/HudsonValley)
Intel Xeon W-2235 CPU 3.80GHz (Max: 3.79GHz), 1 CPU, 12 logical and 6 physical cores
.NET SDK 10.0.100-preview.5.25277.114
  [Host]     : .NET 10.0.0 (10.0.25.27814), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
  Job-QMWUGV : .NET 10.0.0 (42.42.42.42424), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
  Job-HFWDAV : .NET 10.0.0 (10.0.25.27814), X64 RyuJIT AVX-512F+CD+BW+DQ+VL

Runtime=.NET 10.0  

```
| Method                   | Job        | Toolchain | Mean              | Error          | StdDev         | Ratio | RatioSD |
|------------------------- |----------- |---------- |------------------:|---------------:|---------------:|------:|--------:|
| UncontendedHeaderLock    | Job-QMWUGV | CoreRun   |          16.59 ns |       0.235 ns |       0.219 ns |  1.09 |    0.02 |
| UncontendedHeaderLock    | Job-HFWDAV | net10.0   |          15.20 ns |       0.284 ns |       0.266 ns |  1.00 |    0.02 |
|                          |            |           |                   |                |                |       |         |
| UncontendedSyncBlockLock | Job-QMWUGV | CoreRun   |         675.72 ns |      13.556 ns |      19.870 ns |  1.25 |    0.04 |
| UncontendedSyncBlockLock | Job-HFWDAV | net10.0   |         540.03 ns |      10.219 ns |       9.059 ns |  1.00 |    0.02 |
|                          |            |           |                   |                |                |       |         |
| UncontendedUpgradedLock  | Job-QMWUGV | CoreRun   |         703.19 ns |      14.039 ns |      17.241 ns |  1.09 |    0.04 |
| UncontendedUpgradedLock  | Job-HFWDAV | net10.0   |         647.25 ns |      12.811 ns |      16.658 ns |  1.00 |    0.04 |
|                          |            |           |                   |                |                |       |         |
| ContendedLock            | Job-QMWUGV | CoreRun   | 110,667,796.00 ns | 515,885.230 ns | 482,559.364 ns |  1.00 |    0.01 |
| ContendedLock            | Job-HFWDAV | net10.0   | 110,675,577.14 ns | 534,690.047 ns | 473,988.800 ns |  1.00 |    0.01 |

jkoritzinsky committed:
Remove remnants of TBI from the runtime (it's not needed there any more)

Restore FCalls for fast acquire/release and add a true "do only the fastest fast path" path to get back to the expected perf.

Add back contention spinning for the "thin-lock one shot" case

Add extra-fast path for TryEnter

Precreate lock when we're assigning an entry from managed

Do the spinning path without upgrading, refactor how we get the lock object to reduce header mainpulation in managed code

Refactor Release to not call into the sync table from ObjectHeader

Handle the managed thread ID within the runtime for the header lock cases.

Fix handling the upgraded-while-locked case

Move to a simpler model for Lock object lookup

Clean up the Lock shared code

For CoreCLR, initialize the ThreadId type used by System.Threading.Lock as a special thread-static, so we never need to check for initialization in the "upgrade" case.

jkoritzinsky commented:
Updated perf results here (including a better benchmark for the sync-block lock case):

<details>
<summary>Object header lock</summary>

```

BenchmarkDotNet v0.15.2, Windows 11 (10.0.26100.4946/24H2/2024Update/HudsonValley)
Intel Xeon W-2235 CPU 3.80GHz (Max: 3.79GHz), 1 CPU, 12 logical and 6 physical cores
.NET SDK 10.0.100-preview.5.25277.114
  [Host]     : .NET 10.0.0 (10.0.25.27814), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
  DefaultJob : .NET 10.0.0 (10.0.25.27814), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
  Job-QMWUGV : .NET 10.0.0 (42.42.42.42424), X64 RyuJIT AVX-512F+CD+BW+DQ+VL


```
| Method      | Job        | Toolchain | Mean     | Error    | StdDev   | Ratio | RatioSD |
|------------ |----------- |---------- |---------:|---------:|---------:|------:|--------:|
| Uncontended | DefaultJob | Default   | 16.72 ns | 0.321 ns | 0.300 ns |  1.00 |    0.02 |
| Uncontended | Job-QMWUGV | CoreRun   | 16.45 ns | 0.199 ns | 0.177 ns |  0.98 |    0.02 |


</details>

<details>
<summary>Sync Block Lock</summary>

```

BenchmarkDotNet v0.15.2, Windows 11 (10.0.26100.4946/24H2/2024Update/HudsonValley)
Intel Xeon W-2235 CPU 3.80GHz (Max: 3.79GHz), 1 CPU, 12 logical and 6 physical cores
.NET SDK 10.0.100-preview.5.25277.114
  [Host]     : .NET 10.0.0 (10.0.25.27814), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
  DefaultJob : .NET 10.0.0 (10.0.25.27814), X64 RyuJIT AVX-512F+CD+BW+DQ+VL
  Job-QMWUGV : .NET 10.0.0 (42.42.42.42424), X64 RyuJIT AVX-512F+CD+BW+DQ+VL


```
| Method      | Job        | Toolchain | Mean     | Error    | StdDev   | Ratio |
|------------ |----------- |---------- |---------:|---------:|---------:|------:|
| Uncontended | DefaultJob | Default   | 15.05 ns | 0.111 ns | 0.099 ns |  1.00 |
| Uncontended | Job-QMWUGV | CoreRun   | 17.28 ns | 0.213 ns | 0.200 ns |  1.15 |


</details>

Still trying to figure out how to get the sync-block uncontended lock case faster. Might need to move away from the object handle to do that.

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into monitor-managed

jkoritzinsky committed:
Exit the thin-lock spinning when there's a sync-block index added.

jkoritzinsky committed:
Switch debug check to use TryGetLockInfo

jkoritzinsky committed:
Construct the lock object in managed code so we can hit the optimized alloc path during the upgrade.

jkoritzinsky committed:
Don't force EnterSlow to not be inlined.

jkoritzinsky committed:
Manually inline away the *Slow variants

jkoritzinsky committed:
Remove SyncTable class. It isn't worth having a separate type at this point

jkoritzinsky commented:
I've improved the uncontended thick-lock case to only 1.5ns slower and I've exhausted all of my ideas.

| Method      | Job        | Toolchain | Mean     | Error    | StdDev   | Ratio | RatioSD |
|------------ |----------- |---------- |---------:|---------:|---------:|------:|--------:|
| Uncontended | DefaultJob | Default   | 15.17 ns | 0.328 ns | 0.291 ns |  1.00 |    0.03 |
| Uncontended | Job-QMWUGV | CoreRun   | 16.88 ns | 0.155 ns | 0.137 ns |  1.11 |    0.02 |

At this point it's as fast as the uncontended thin-lock case (within measurement error).

Can I get another review pass?

jkoritzinsky committed:
Fix naming issues

MihaZupan commented:
`System.Collections.Concurrent.AddRemoveFromDifferentThreads<Int32>.ConcurrentBag(Size: 2000000)` got stuck spinning one core, looks like at `ObjHeader::PassiveGetSyncBlock`.

jkotas commented:
> System.Collections.Concurrent.AddRemoveFromDifferentThreads<Int32>.ConcurrentBag(Size: 2000000) got stuck spinning one core.

Sounds like a bug that we are "porting" to coreclr now?

jkoritzinsky committed:
Flip condition to skip out on the thin lock spin logic.

jkoritzinsky committed:
Don't try to do a quick-loop for release. Just go to the slow path like how it's done in main.

jkoritzinsky commented:
Looking at the linked issue, I think the NativeAOT failure was due to https://github.com/dotnet/runtime/issues/67805 (linked from https://github.com/dotnet/runtime/issues/66987). I also think that https://github.com/dotnet/runtime/pull/73033 may have contributed to fixing it.

I think the failures from this PR's run of the benchmark were due to bugs in the lock-free algorithms I wrote (same cause as the PR failures).

jkoritzinsky committed:
Rewrite IsEntered implementation to match how Enter and Exit are implemented.

jkoritzinsky committed:
Actually fix the DBI asserts

jkoritzinsky committed:
Try-Enter with a timeout should do the slow path if there's a timeout

jkoritzinsky committed:
Check for syncblk index first. Thread-id is only valid if this bit is false

jkoritzinsky committed:
Unify enum definitions

jkotas commented:
@MihuBot benchmark System.Threading

jkoritzinsky commented:
I ran these benchmarks locally and used the perf team's ResultsComparer tooling (with their recommended 2% threshold) and got the following results:

```
summary:
better: 23, geomean: 1.099
worse: 11, geomean: 1.093
total diff: 34
```

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Threading.Tasks.Tests.Perf_AsyncMethods.Yield                             |      1.43 |           157.16 |           224.16 |         |
| System.Threading.Channels.Tests.UnboundedChannelPerfTests.TryWriteThenTryRead    |      1.14 |            29.64 |            33.64 |         |
| System.Collections.Concurrent.AddRemoveFromDifferentThreads<String>.ConcurrentSt |      1.11 |      89349750.00 |      98731550.00 | several?|
| System.Threading.Tasks.ValueTaskPerfTest.CreateAndAwait_FromResult_ConfigureAwai |      1.07 |             5.32 |             5.70 |         |
| System.Threading.Tests.Perf_Timer.AsynchronousContention                         |      1.07 |    4311523100.00 |    4596094950.00 |         |
| System.Threading.Tasks.ValueTaskPerfTest.CreateAndAwait_FromCompletedValueTaskSo |      1.06 |             8.60 |             9.14 |         |
| System.Threading.Channels.Tests.BoundedChannelPerfTests.ReadAsyncThenWriteAsync  |      1.06 |            60.48 |            63.94 | bimodal |
| System.Threading.Tasks.Tests.Perf_AsyncMethods.SingleYieldMethodInvocation       |      1.05 |           496.56 |           519.55 |         |
| System.Threading.Tasks.ValueTaskPerfTest.Await_FromCompletedTask                 |      1.04 |            12.26 |            12.80 |         |
| System.Threading.Channels.Tests.SpscUnboundedChannelPerfTests.WriteAsyncThenRead |      1.04 |            31.39 |            32.50 |         |
| System.Threading.Channels.Tests.BoundedChannelPerfTests.WriteAsyncThenReadAsync  |      1.03 |            46.94 |            48.23 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| -------- |
| System.Collections.Concurrent.IsEmpty<String>.Bag(Size: 512)                     |      1.29 |             7.57 |             5.87 |         |
| System.Collections.Concurrent.IsEmpty<String>.Queue(Size: 0)                     |      1.22 |             2.31 |             1.90 |         |
| System.Collections.Concurrent.IsEmpty<String>.Dictionary(Size: 512)              |      1.19 |             3.11 |             2.61 | bimodal |
| System.Threading.Tests.Perf_ThreadStatic.SetThreadStatic                         |      1.18 |             2.20 |             1.86 |         |
| System.Threading.Tasks.ValueTaskPerfTest.Copy_PassAsArgumentAndReturn_FromResult |      1.16 |             2.36 |             2.03 |         |
| System.Collections.Concurrent.IsEmpty<String>.Bag(Size: 0)                       |      1.16 |             7.17 |             6.18 |         |
| System.Threading.Tests.Perf_SemaphoreSlim.ReleaseWait                            |      1.11 |            35.54 |            31.97 |         |
| System.Threading.Tasks.Tests.Perf_AsyncMethods.EmptyAsyncMethodInvocation        |      1.10 |             5.17 |             4.70 |         |
| System.Threading.Tests.Perf_SemaphoreSlim.ReleaseWaitAsync_WithCancellationToken |      1.09 |          1570.68 |          1435.42 |         |
| System.Threading.Tests.Perf_Timer.SynchronousContention                          |      1.08 |    3671500800.00 |    3396126000.00 |         |
| System.Threading.Channels.Tests.BoundedChannelPerfTests.PingPong                 |      1.08 |      13669746.15 |      12649100.00 |         |
| System.Threading.Tests.Perf_Monitor.TryEnterExit                                 |      1.08 |            14.16 |            13.16 |         |
| System.Threading.Channels.Tests.SpscUnboundedChannelPerfTests.TryWriteThenTryRea |      1.07 |            21.01 |            19.56 |         |
| System.Threading.Tasks.ValueTaskPerfTest.CreateAndAwait_FromResult               |      1.07 |             6.04 |             5.64 |         |
| System.Threading.Channels.Tests.UnboundedChannelPerfTests.PingPong               |      1.07 |      12876008.33 |      12026150.00 |         |
| System.Threading.Tasks.ValueTaskPerfTest.CreateAndAwait_FromCompletedValueTaskSo |      1.06 |            14.69 |            13.84 |         |
| System.Threading.Tests.Perf_Monitor.EnterExit                                    |      1.05 |            13.85 |            13.13 |         |
| System.Collections.Concurrent.Count<String>.Queue_EnqueueCountDequeue(Size: 512) |      1.05 |            17.20 |            16.32 |         |
| System.Collections.Concurrent.Count<String>.Bag(Size: 512)                       |      1.05 |            32.46 |            30.81 |         |
| System.Collections.Concurrent.Count<Int32>.Bag(Size: 512)                        |      1.04 |            32.17 |            30.88 |         |
| System.Threading.Tests.Perf_Timer.ScheduleManyThenDisposeMany                    |      1.04 |     487751500.00 |     470460600.00 |         |
| System.Threading.Channels.Tests.SpscUnboundedChannelPerfTests.ReadAsyncThenWrite |      1.03 |            61.28 |            59.30 |         |
| System.Collections.Concurrent.Count<String>.Stack(Size: 512)                     |      1.03 |           478.25 |           464.98 |         |

The `Task` and `ValueTask` ones seem unrelated, and the only ones that are consistently slower are the `Channels` ones.

I'm not sure if this is due to different core counts, the fact that MihuBot runs on cloud VMs, or the natural instability of multithreading benchmarks.

jkotas commented:
ArrayWithThread timeout in the last CI run looks related. From `runfo get-helix-payload -j 991554c5-582f-4a5f-a3b9-ad86f7f4d8d9 -w jit64_3 -o c:\helix_payload`:

One thread is waiting for Lock cctor to finish running:
```
ntdll!RtlEnterCriticalSection+0x40 [d:\rs1\minkernel\ntdll\resource.c @ 1766] 
coreclr!CrstBase::Enter+0x171 [D:\a\_work\1\s\src\coreclr\vm\crst.cpp @ 262] 
coreclr!ListLockEntryBase<void *>::FinishDeadlockAwareEnter+0x24 [D:\a\_work\1\s\src\coreclr\vm\listlock.h @ 202] 
coreclr!ListLockEntryBase<void *>::LockHolder::DeadlockAwareAcquire+0x35 [D:\a\_work\1\s\src\coreclr\vm\listlock.h @ 227] 
coreclr!MethodTable::DoRunClassInitThrowing+0x315 [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3708] 
coreclr!MethodTable::CheckRunClassInitThrowing+0x499 [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3852] 
coreclr!DynamicHelperFixup+0x42c [D:\a\_work\1\s\src\coreclr\vm\prestub.cpp @ 3243] 
coreclr!DynamicHelperWorker+0x15e [D:\a\_work\1\s\src\coreclr\vm\prestub.cpp @ 3553] 
coreclr!DelayLoad_Helper+0x7a [D:\a\_work\1\s\src\coreclr\vm\amd64\ExternalMethodFixupThunk.asm @ 59] 
System_Private_CoreLib!System.Threading.Monitor.<GetLockObject>g__GetLockObjectFallback|0_0(System.Object)+0x37 [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs @ 39] 
System_Private_CoreLib!System.Threading.Monitor.Exit(System.Object)+0x28 [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs @ 154] 
System_Private_CoreLib!System.AppContext.GetData(System.String)+0x61 [/_/src/libraries/System.Private.CoreLib/src/System/AppContext.cs @ 49] 
System_Private_CoreLib!System.AppContext.TryGetSwitch(System.String, Boolean ByRef)+0x39 [/_/src/libraries/System.Private.CoreLib/src/System/AppContext.cs @ 160] 
System_Private_CoreLib!System.Resources.ResourceReader..cctor()+0x1b [/_/src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs @ 102] 
coreclr!CallDescrWorkerInternal+0x83 [D:\a\_work\1\s\src\coreclr\vm\amd64\CallDescrWorkerAMD64.asm @ 74] 
coreclr!`DispatchCallDebuggerWrapper'::`3'::__Body::Run+0x85 [D:\a\_work\1\s\src\coreclr\vm\callhelpers.cpp @ 139] 
coreclr!DispatchCallDebuggerWrapper+0x22 [D:\a\_work\1\s\src\coreclr\vm\callhelpers.cpp @ 143] 
coreclr!DispatchCallSimple+0x180 [D:\a\_work\1\s\src\coreclr\vm\callhelpers.cpp @ 236] 
coreclr!MethodTable::RunClassInitEx+0x475 [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3548] 
coreclr!MethodTable::DoRunClassInitThrowing+0x3db [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3727] 
coreclr!MethodTable::CheckRunClassInitThrowing+0x499 [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3852] 
coreclr!ReflectionInvocation_RunClassConstructor+0x146 [D:\a\_work\1\s\src\coreclr\vm\reflectioninvocation.cpp @ 1230] 
System_Private_CoreLib!System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor(System.RuntimeTypeHandle)+0x8f [/_/src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs @ 174] 
System_Private_CoreLib!System.SR.InternalGetResourceString(System.String)+0x37c [/_/src/libraries/System.Private.CoreLib/src/System/SR.cs @ 86] 
System_Private_CoreLib!System.SR.GetResourceString(System.String)+0x29 [/_/src/libraries/Common/src/System/SR.cs @ 41] 
System_Private_CoreLib!System.IndexOutOfRangeException..ctor()+0x18 [/_/src/libraries/System.Private.CoreLib/src/System/IndexOutOfRangeException.cs @ 17] 
System_Private_CoreLib!Internal.Runtime.CompilerHelpers.ThrowHelpers.ThrowIndexOutOfRangeException()+0x20 [/_/src/libraries/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs @ 46] 
ArrayWithThread_o!ArrayWithThread.Test.Test1(Int32 ByRef)+0x6a
```

Thread running Lock cctor is stuck because it wants to acquire lock that the first thread is holding:

coreclr!Thread::DoAppropriateAptStateWait+0x119 [D:\a\_work\1\s\src\coreclr\vm\threads.cpp @ 2938] 
coreclr!Thread::DoAppropriateWait+0x640 [D:\a\_work\1\s\src\coreclr\vm\threads.cpp @ 3109] 
coreclr!WaitHandle_WaitOneCore+0x191 [D:\a\_work\1\s\src\coreclr\vm\comwaithandle.cpp @ 33] 
System_Private_CoreLib!System.Threading.WaitHandle.WaitOneNoCheck(Int32, Boolean, System.Object, WaitHandleWaitSourceMap)+0x2b6 [/_/src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.cs @ 198] 
System_Private_CoreLib!System.Threading.Lock.TryEnterSlow(Int32, ThreadId)+0x5e1 [/_/src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs @ 524] 
System_Private_CoreLib!System.Threading.Lock.Enter()+0xd4 [/_/src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs @ 74] 
System_Private_CoreLib!System.Threading.Monitor.Enter(System.Object, Boolean ByRef)+0x32 [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs @ 85] 
System_Private_CoreLib!System.AppContext.GetData(System.String)+0x3e [/_/src/libraries/System.Private.CoreLib/src/System/AppContext.cs @ 47] 
System_Private_CoreLib!System.AppContextConfigHelper.GetInt16Config(System.String, Int16, Boolean)+0x1b [/_/src/libraries/System.Private.CoreLib/src/System/AppContextConfigHelper.cs @ 161] 
System_Private_CoreLib!System.Threading.Lock..cctor()+0x14 [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Lock.CoreCLR.cs @ 11] 
coreclr!CallDescrWorkerInternal+0x83 [D:\a\_work\1\s\src\coreclr\vm\amd64\CallDescrWorkerAMD64.asm @ 74] 
coreclr!`DispatchCallDebuggerWrapper'::`3'::__Body::Run+0x85 [D:\a\_work\1\s\src\coreclr\vm\callhelpers.cpp @ 139] 
coreclr!DispatchCallDebuggerWrapper+0x22 [D:\a\_work\1\s\src\coreclr\vm\callhelpers.cpp @ 143] 
coreclr!DispatchCallSimple+0x180 [D:\a\_work\1\s\src\coreclr\vm\callhelpers.cpp @ 236] 
coreclr!MethodTable::RunClassInitEx+0x475 [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3548] 
coreclr!MethodTable::DoRunClassInitThrowing+0x3db [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3727] 
coreclr!MethodTable::CheckRunClassInitThrowing+0x499 [D:\a\_work\1\s\src\coreclr\vm\methodtable.cpp @ 3852] 
coreclr!DynamicHelperFixup+0x42c [D:\a\_work\1\s\src\coreclr\vm\prestub.cpp @ 3243] 
coreclr!DynamicHelperWorker+0x15e [D:\a\_work\1\s\src\coreclr\vm\prestub.cpp @ 3553] 
coreclr!DelayLoad_Helper+0x7a [D:\a\_work\1\s\src\coreclr\vm\amd64\ExternalMethodFixupThunk.asm @ 59] 
System_Private_CoreLib!System.Threading.Monitor.<GetLockObject>g__GetLockObjectFallback|0_0(System.Object)+0x37 [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs @ 39] 
System_Private_CoreLib!System.Threading.Monitor+<>c.<.cctor>b__26_0(System.Object)+0x1b [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs @ 57] 
System_Private_CoreLib!System.Runtime.CompilerServices.ConditionalWeakTable`2[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]].GetOrAdd(System.__Canon, System.Func`2<System.__Canon,System.__Canon>)+0x3b [/_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs @ 247] 
System_Private_CoreLib!System.Threading.Monitor.PulseAll(System.Object)+0xf [/_/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs @ 191]

jkoritzinsky commented:
I'll port over that logic from NAOT and add a comment block about how we can hit it with CoreCLR (as the exact scenario we can hit it is a little different)

jkoritzinsky committed:
Port over the lazy statics initialization to avoid cctor deadlocks.

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into monitor-managed

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into monitor-managed

jkotas committed:
Merge branch 'main' into monitor-managed

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs:
```diff
@@ -22,329 +21,185 @@ namespace System.Threading
-        private static partial bool Wait(ObjectHandleOnStack obj, int millisecondsTimeout);
+        #endregion
+
+        #region Public Wait/Pulse methods
```
Everything from here still the end of the file looks identical to NAOT. Move to shared?

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs:
```diff
@@ -22,329 +21,185 @@ namespace System.Threading
+            GetLockObject(obj).Exit();
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Ah, I guess it is to prevent recursive inlining of `ObjectHeader.IsAcquired`. It may use comment.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs:
```diff
@@ -22,329 +21,185 @@ namespace System.Threading
         {
+            // we are inlining lockTaken check as the check is likely be optimized away
             if (lockTaken)
-                ThrowLockTakenException();
```
Keep `ThrowLockTakenException` throw helper to make this more likely to be inlined and prevent eating JIT inlining budgets?

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/ThreadBlockingInfo.cs:
```diff
@@ -185,8 +128,10 @@ public void Dispose()
+            [Obsolete("Represents native Monitor locks, which have been removed", error: true)]
+            MonitorLock, // maps to DebugBlockingItemType::Legacy_DebugBlock_MonitorCriticalSection in coreclr
+            [Obsolete("Represents native Monitor waits, which have been removed", error: true)]
+            MonitorWait, // maps to DebugBlockingItemType::Legacy_DebugBlock_MonitorEvent in coreclr
```
```suggestion
            [Obsolete("Represents native Monitor locks, which have been removed", error: true)]
            MonitorLock,
            [Obsolete("Represents native Monitor waits, which have been removed", error: true)]
            MonitorWait,
```
This PR is deleting Legacy_DebugBlock_MonitorCriticalSection/Legacy_DebugBlock_MonitorEvent so there is no useful in referencing them from these comments.

jkotas commented on src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.Shared.xml:
```diff
@@ -62,6 +62,11 @@
     </type>
+
+    <type fullname="System.Threading.ThreadBlockingInfo">
+      <property name="LockOwnerManagedThreadId" />
```
Was the debugger feature broken without this?

jkoritzinsky committed:
Unify System.Threading.Lock type implementation and define split ManagedThreadId definitions (the last place that differs between runtimes)

jkoritzinsky committed:
Remove PulseAllHelper (matches dotnet/runtime#121132)

jkoritzinsky commented on src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs:
```diff
@@ -22,329 +21,185 @@ namespace System.Threading
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static bool TryEnter_Slowpath(object obj, int timeout)
+        #region Public Enter/Exit methods
+        public static void Enter(object obj)
```
I had a microbenchrmark sittting around that has the generated assembly (with Enter inlined, but easily identifiable):

```
    [Benchmark]
    public void Uncontended()
    {
        object o = new();
        Monitor.Enter(o);
        Monitor.Exit(o);
    }
```

Here's the assembly (annotated):

```asm
       push      rbx
       sub       rsp,20
       mov       rcx,offset MT_System.Object
       call      CORINFO_HELP_NEWSFAST
		# Start of Monitor.Enter
       mov       rbx,rax
       mov       rcx,rbx
       call      00007FFBC8C4EC80 # Call ObjectHeader.AcquireInternal
       cmp       eax,1
       je        short M00_L02
M00_L00:
       test      eax,eax
       jne       short M00_L03
M00_L01:
      # This IG is not part of Monitor.Enter
       mov       rcx,rbx
       add       rsp,20
       pop       rbx
       jmp       qword ptr [7FFB68FD6868]; System.Threading.Monitor.Exit(System.Object)
M00_L02:
		# 
       mov       rcx,rbx
       call      qword ptr [7FFB693BEE50] # Call ObjectHeader.TryAcquireThinLockSpin
       jmp       short M00_L00
M00_L03:
       # Slow path (GetLockObject(obj).Enter())
       mov       rcx,rbx
       call      qword ptr [7FFB693BEE80]
       mov       rcx,rax
       cmp       [rcx],ecx
       call      qword ptr [7FFB693BF2E8]
       jmp       short M00_L01
```

So, we inline the fast path (acquiring the header bit) and don't inline the header bit spinning or the fat lock path.

This is very similar to the existing Monitor code in main (in fact, the only way I was able to get anywhere near similar performance was to match the inlining characteristics and code structure for the thin-lock case).

jkoritzinsky committed:
PR feedback.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs:
```diff
@@ -22,329 +22,145 @@ namespace System.Threading
 
-        // Used to implement synchronized methods on non Windows-X86 architectures
-        internal static void ExitIfLockTaken(object obj, ref bool lockTaken)
+        public static void Enter(object obj, ref bool lockTaken)
```
Can be moved to shared and marked AggresiveInlining to match NAOT

jkotas commented on src/coreclr/vm/jitinterface.h:
```diff
@@ -171,7 +170,6 @@ EXTERN_C FCDECL2(Object*, RhpNewArrayMaybeFrozen, MethodTable* pMT, INT_PTR size
 
 EXTERN_C FCDECL2(void, JITutil_MonReliableEnter, Object* obj, BYTE* pbLockTaken);
 EXTERN_C FCDECL3(void, JITutil_MonTryEnter, Object* obj, INT32 timeOut, BYTE* pbLockTaken);
```
```suggestion
```
Unused?

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/Lock.cs:
```diff
@@ -1265,11 +1490,37 @@ public static bool TrySetIsWaiterSignaledToWake(Lock lockObj, State state)
         }
 
+
+        internal struct ThreadId
```
Is this wrapper still needed now that thread id is managed thread id? Also, can we make it `int` since that's what the managed thread ID is in the public surface?

jkoritzinsky replied:
We can remove this wrapper.

jkoritzinsky committed:
PR feedback

jkoritzinsky committed:
Remove ThreadId abstraction

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Threading/SynchronizedMethodHelpers.CoreCLR.cs:
```diff
@@ -0,0 +1,50 @@
+    /// </summary>
+    internal static class SynchronizedMethodHelpers
+    {
+        private static void MonitorEnter(object obj, ref bool lockTaken)
```
We have been moving away from the native AOT factoring with JIT helpers on the side under `Internal\Runtime\CompilerHelpers\*`, and placing the JIT helpers into their natural homes instead. For example, we used to have `Internal\Runtime\CompilerHelpers\MathHelpers.cs` that is gone now and merged into System.Math. It is better for CoreCLR (fewer types loaded) and the code is easier to understand when everything related is in the same file.

I would move Monitor.Enter/Exit helpers to Monitor, both here and in the NativeAOT version.

jkotas replied:
(Give them distinct name like `SynchronizedMethodEnter/Exit` so that there are no collisions with the regular Enter/Exit.)

jkotas reviewed (approved):
LGTM (modulo one last nit)

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Threading/Monitor.cs:
```diff
@@ -36,5 +39,88 @@ public static bool Wait(object obj, int millisecondsTimeout, bool exitContext)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void TryEnter(object obj, int millisecondsTimeout, ref bool lockTaken)
+        {
+            if (lockTaken)
```
```suggestion
            // Aggressively inline lockTaken check as it is likely to be optimized away
            if (lockTaken)
```

jkoritzinsky committed:
Remove SynchronizedMethodHelpers, fix up NativeAOT build.

jkoritzinsky committed:
Add missing parameter

jkoritzinsky committed:
Add missing cast
